docs/proposals/metric-tab-and-metric-puller.md

```diff
@@ -0,0 +1,134 @@
+
+## Proposal
+
+#####ArgoCD UI:
+Enhancing ArgoCD UI to support the Metrics tab on the resource page. This Tab will be enabled and disabled using the feature flag. 
+This Metrics component will make an API call to render the metrics dashboard from the new Metrics Server.
+
+#####ArgoCD Server:
+Enhancing ArgoCD Server to have metrics puller and will expose the APIs to provide the Dashboard and Metrics information. Metrics Puller will read the Dashboard, Graph, and metrics provider configuration from Configmap.  
+
+#####Authentication:
+No need for separate implementation
+
+#####Defining Metrics page:
+The metrics page is fully configurable. UI will call API to get the dashboard layout and graph information from the server. User
+
+#####Enable/Disable the feature:
+The server will return HTTP resource code:405  if the feature is not enabled. UI will show/hide the metrics tab.
+
+
+### Use cases
+1. User can compare KPIs for rollout releases  quickly.
+2. User can review the service performance.
+
+### Implementation Details/Notes/Constraints [optional]
+
+##### Metrics Tap:
+This feature is going to be first class citizen in ArgoCDUI. Adding Metrics Component in Resource page. Implement Metrics
+Service  to make API call to query the Dashboard configuration and Graph data.
+
+#####Metrics Puller:
+Metrics Puller module will be implemented in ArgoCD Server. ArgoCD API server to expose the APIs for Metrics Dashboard and Graph data.
+
+Metric Dashboard and Graph query can be fully configurable.  All Dashboard and Graph query  will be configure in Configmap. 
+Metric puller will read the appropriate dashboard configuration based on application/resource kind. Metrics puller will read the graph
+query from configmap and execute it in configured metrics provider(`prometheus`).
+
+#####Architecture:
+![Architecture](images/ArgoCD_Metrics_Proposal.jpeg)
+
+
+New APIs:
+`/api/metrics/:application/:cluster/:group/:kind/dashboard → Get Dashboard config`
+`/api/metrics/:application/:cluster/:group/:kind/:row/:graph→ Get Metrics for Graph`
+
+Dashboard and Graph Configuration:
+
+Application, Dashboard and cluster has `default` element.  Default Application will be return if there is not special config for application.
+Metrics puller will execute query on default cluster if there is no config for the cluster.
+
+```yaml
+{
+  "prometheus": {
+    "applications": [
+      {
+        "name": "default",
+        "default": true,
+        "dashboards": [
+          {
+            "group": "v1",
+            "kind": "pod",
+            "rows": [
+              {
+                "name": "HTTP",
+                "graphs": [
+                  {
+                    "name": "http_duration",
+                    "title": "HTTP duration",
+                    "graphType": "pie",
+                    "duration": "1d",
+                    "queryExpression": "rate(container_cpu_user_seconds_total{pod=\"{{.name}}\", namespace={{.namespace}}}[30s])*100"
```

Comment:
> A few questions regarding handling prometheus queries:
> 
> 1. How the prometheus response object looks like for this query?
- We can provide this.
> 2. Does it have all the necessary information to be able to draw a diagram in ArgoCD UI? Or we would need to make multiple calls?
- Yes each call responds with the needed information per object type per metric type. For example one call for PODs/CPU and another for PODs/MEM.
> 3. ArgoCD UI will keep sending requests to API server to get refreshed diagram data?
- Yes UI makes subsequent requests as the props data in ArgoCD get refreshed. This can be tuned if there are any concerns (ie. ignore new props, or only update after a certain timelapse)